client-runtime/src/main/java/com/microsoft/rest/v2/util/FlowableUtil.java

```diff
@@ -70,6 +72,73 @@ public static boolean isFlowableByteBuffer(TypeToken entityTypeToken) {
 
         });
     }
+    private static final class CountingFlowable extends Flowable<ByteBuffer> {
+        final Flowable<ByteBuffer> source;
+        final long bytesExpected;
+        CountingFlowable(Flowable<ByteBuffer> source, long bytesExpected) {
+            if (bytesExpected < 0) {
+                throw new IllegalArgumentException("bytesExpected must be non-negative, but was " + bytesExpected);
+            }
+
+            this.source = Objects.requireNonNull(source, "source is null");
+            this.bytesExpected = bytesExpected;
+        }
+
+        @Override
+        protected void subscribeActual(Subscriber<? super ByteBuffer> observer) {
+            source.subscribe(new Subscriber<ByteBuffer>() {
+                Subscription subscription;
+                long bytesRead = 0;
+
+                @Override
+                public void onSubscribe(Subscription s) {
+                    subscription = s;
+                    observer.onSubscribe(s);
+                }
+
+                @Override
+                public void onNext(ByteBuffer byteBuffer) {
+                    long newBytesRead = bytesRead + byteBuffer.remaining();
```

Comment:
Because onNext can call onError you need protection here. The reason is if onNext routes to onError and the next event is say onComplete it is against the contract to emit events after the terminal event. What you do to prevent it is:

* declare non-volatile field `boolean done;`
* in onNext:
```java
if (done) {
  return;
}
...
```
* in onError:
```java
if (done) {
  RxJavaPlugins.onError(t);
  return;
}
done = true;
...
```
* in onComplete:
```java
if (done) {
  return;
}
done = true;
...
```